drivers/platform/stm32/stm32_tdm.c

```diff
@@ -0,0 +1,219 @@
+ * @param param - The structure that contains the TDM parameters.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t stm32_tdm_init(struct tdm_desc **desc,
+		       const struct tdm_init_param *param)
+{
+	int32_t ret;
+	uint32_t tmp;
+	struct tdm_desc	*tdm_desc;
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	if (param->mode > TDM_SLAVE_RX)
+		return -EINVAL;
+
+	tdm_desc = (struct tdm_desc *)calloc(1, sizeof(*tdm_desc));
+	if (!tdm_desc)
+		return -ENOMEM;
+
+	struct stm32_tdm_desc *tdesc;
+	struct stm32_tdm_init_param *tinit;
+
+	tdesc = (stm32_tdm_desc*)calloc(1,sizeof(stm32_tdm_desc));
+	if (!tdesc) {
+		ret = -ENOMEM;
+		goto error;
+	}
+
+	tdm_desc->extra = tdesc;
+	tinit = param->extra;
+
+	tdesc->hsai.Instance = tinit->base;
+	tdesc->hsai.Init.Protocol = SAI_FREE_PROTOCOL;
+	switch(param->mode) {
+	case TDM_MASTER_TX:
+	case TDM_MASTER_RX:
+	case TDM_SLAVE_TX:
+		// TODO: implement master tx/rx and slave tx
+		ret = -ENOTSUP;
+		goto error;
+	case TDM_SLAVE_RX:
+		tmp = SAI_MODESLAVE_RX;
+		break;
+	default:
+		ret = -EINVAL;
+		break;
+	};
+	tdesc->hsai.Init.AudioMode = tmp;
+	switch(param->data_size) {
+	case 8:
+		tmp = SAI_DATASIZE_8;
+		break;
+	case 10:
+		tmp = SAI_DATASIZE_10;
+		break;
+	case 16:
+		tmp = SAI_DATASIZE_16;
+		break;
+	case 20:
+		tmp = SAI_DATASIZE_20;
+		break;
+	case 24:
+		tmp = SAI_DATASIZE_24;
+		break;
+	case 32:
+		tmp = SAI_DATASIZE_32;
+		break;
+	default:
+		ret = -ENOTSUP;
+		goto error;
+		break;
+	};
+	tdesc->hsai.Init.DataSize = tmp;
+	tdesc->hsai.Init.FirstBit = param->data_lsb_first ? SAI_FIRSTBIT_LSB :
+				    SAI_FIRSTBIT_MSB;
+	tdesc->hsai.Init.ClockStrobing = param->rising_edge_sampling ?
+					 SAI_CLOCKSTROBING_RISINGEDGE : SAI_CLOCKSTROBING_FALLINGEDGE;
+	tdesc->hsai.Init.Synchro = SAI_ASYNCHRONOUS;
+	tdesc->hsai.Init.OutputDrive = SAI_OUTPUTDRIVE_DISABLE;
+	tdesc->hsai.Init.FIFOThreshold = SAI_FIFOTHRESHOLD_EMPTY;
+	tdesc->hsai.Init.SynchroExt = SAI_SYNCEXT_DISABLE;
+	tdesc->hsai.Init.MonoStereoMode = SAI_STEREOMODE;
+	tdesc->hsai.Init.CompandingMode = SAI_NOCOMPANDING;
+	tdesc->hsai.Init.TriState = SAI_OUTPUT_NOTRELEASED;
+	tdesc->hsai.FrameInit.FrameLength = (param->data_size + param->data_offset) *
+					    param->slots_per_frame;
+	tdesc->hsai.FrameInit.ActiveFrameLength = param->fs_active_length;
+	tdesc->hsai.FrameInit.FSDefinition = SAI_FS_STARTFRAME;
+	tdesc->hsai.FrameInit.FSPolarity = param->fs_active_low ? SAI_FS_ACTIVE_LOW :
+					   SAI_FS_ACTIVE_HIGH;
+	tdesc->hsai.FrameInit.FSOffset = param->fs_lastbit ? SAI_FS_BEFOREFIRSTBIT :
+					 SAI_FS_FIRSTBIT;
+	tdesc->hsai.SlotInit.FirstBitOffset = param->data_offset;
+	tdesc->hsai.SlotInit.SlotSize = SAI_SLOTSIZE_DATASIZE;
+	tdesc->hsai.SlotInit.SlotNumber = param->slots_per_frame;
+	tdesc->hsai.SlotInit.SlotActive = (1u << param->slots_per_frame) - 1u;
+	if (HAL_SAI_Init(&tdesc->hsai) != HAL_OK) {
+		ret = -EIO;
+		goto error;
+	}
+
+	*desc = tdm_desc;
+
+	return SUCCESS;
```

Comment:
Is `SUCCESS` defined in `errno.h` ? 
Otherwise I would include `error.h` in this file.

Then you can replace all `return 0;` with `return SUCCESS` (See line 185 and 206)